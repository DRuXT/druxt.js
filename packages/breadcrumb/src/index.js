import { DruxtBreadcrumbModule } from './module'

/**
 * The Nuxt.js module function.
 *
 * Installs the module functionality in a Nuxt.js frontend.
 *
 * @type {Function}
 * @exports default
 * @name DruxtBreadcrumbModule
 * @see {@link ./module|DruxtBreadcrumbModule}
 */
export default DruxtBreadcrumbModule

/**
 * Vue.js Mixin.
 *
 * @type {object}
 * @exports DruxtBreadcrumbMixin
 * @see {@link ./mixins/breadcrumb|DruxtBreadcrumbMixin}
 */
 export { DruxtBreadcrumbMixin } from './mixins/breadcrumb'
