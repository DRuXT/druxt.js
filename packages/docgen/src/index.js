import consola from 'consola'
import dmd from 'dmd'
import fs from 'fs'
import globby from 'globby'
import jsdoc2md from 'jsdoc-to-markdown'
import mkdirp from 'mkdirp'
import ncp from 'ncp'
import path from 'path'
import vueDocs from 'vue-docgen-api'

const cwd = path.join(__dirname, '..')

/**
 * DruxtDocgen class.
 */
class DruxtDocgen {
  constructor() {
    this.components = []
    this.destination = 'docs/content'
  }

  /**
   * Generate documentation for source JS and Vue files.
   */
  async generateDocs() {
    consola.info('Generating docs')

    /**
     * Generate API documents.
     */
    await this.generateApiDocs()

    /**
     * Generate list of packages.
     */
    await this.generatePackageList()

    /**
     * Generate list of components.
     */
    await this.generateComponentsList()

    /**
     * Copy external files to the generated files.
     */
    await this.copyFiles()
  }

  /**
   * Copy external files to the generated files.
   */
  async copyFiles() {
    consola.info('Copying files')

    // Changelogs.
    const changelogs = await globby('packages/*/CHANGELOG.md')
    for (const file of changelogs) {
      const module = file.split('/')[1]
      const destination = `${this.destination}/api/packages/${module}/CHANGELOG.md`
      const options = {
        transform: (read, write) => {
          // Append frontmatter
          write.write("---\ntitle: Release notes\n---\n\n")
          read.pipe(write)
        }
      }
      await ncp(file, destination, options, () => {})
    }

    // Contributing guide.
    await ncp('CONTRIBUTING.md', `${this.destination}/guide/CONTRIBUTING.md`)
  }

  /**
   * Generate API documents.
   */
  async generateApiDocs() {
    consola.info('Generating API documents')

    const files = await globby([
      // Javascript.
      'packages/*/src/**/*.js',
      // Vue.js.
      'packages/*/src/components/**/*.vue',
      // Exclude fixtures, mocks and tests.
      '!**/__*__/**/*',
    ])

    for (const file of files) {
      // Get JSDoc template data from file.
      const templateData = jsdoc2md.getTemplateDataSync({
        configure: path.resolve(__dirname, '../jsdoc.json'),
        files: file,
      })

      // Process file based on extension.
      if (file.match(/\.js$/)) {
        await this.processJs(file, templateData)
      }
      else {
        this.components.push({ file, templateData })
        await this.processVue(file, templateData)
      }

      this.writeTemplateData(file, templateData)
    }
  }

  /**
   * Generate list of components.
   */
  async generateComponentsList() {
    consola.info('Generating components list')

    const components = this.components
      .filter((o) => o && o.file.split('/').length === 5 && o.templateData[0])
      .map(({ file, templateData }) =>
        `## ${templateData[0].name}

${templateData[0].description}

[Read the docs](/api/${file.replace('src/', '').replace('.vue', '')})
`)

    const destination = this.destination + '/api/components.md'
    mkdirp.sync(path.dirname(destination))

    fs.writeFileSync(destination, `---
title: Components
---
${components.join('\n* * *\n\n')}

* * *

_Note: The contents of this file where automatically generated by the [Druxt Docgen](https://github.com/druxt/druxt.js/tree/develop/packages/docgen)._
`)
  }

  /**
   * Generate list of packages.
   */
  async generatePackageList() {
    consola.info('Generating package list')

    const packageFiles = await globby('packages/*/package.json')
    const packages = packageFiles
      // Load package data.
      .map((file) => ({
        ...require(path.resolve(__dirname, '../../..', file)),
        __dir: file.split('/')[1],
      }))
      // Filter and sort data
      .filter((o) => !o.private)
      .sort((a, b) => a.name < b.name ? -1 : 1)
      // Process package data.
      .map((pkg) =>
        `## ${pkg.name}

<div class="text-sm">

Version: [${pkg.version}](/api/packages/${pkg.__dir}/CHANGELOG)

</div>

${pkg.description}

[Read the docs](/api/packages/${pkg.__dir}/)
`)

    const destination = this.destination + '/api/README.md'
    mkdirp.sync(path.dirname(destination))

    fs.writeFileSync(destination, `---
title: Packages
---

${packages.join('\n* * *\n\n')}

* * *

_Note: The contents of this file where automatically generated by the [Druxt Docgen](https://github.com/druxt/druxt.js/tree/develop/packages/docgen)._
`)
  }

  /**
   * Process Javascript files.
   */
  async processJs(file, templateData) {
    templateData.map(item => {
      // Vuex state scope fix.
      if (item.name === 'state' && item.scope === 'inner') {
        delete item.scope
      }

      // Vuex mutations/@mutator tag suport.
      if (item.mutators) {
        item.description = item.mutators[0].description
        item.kind = 'method'
        item.scope = 'mutation'
      }
    })
  }

  /**
   * Process Vue.js files.
   */
  async processVue(file, templateData) {
    // Get data from vue-docgen-api.
    const data = await vueDocs.parse(file)

    templateData.map(item => {
      const parts = item.id.split('.')
      if (item.memberof === null && parts.length > 1) {
        parts.pop()
        item.memberof = parts.join('.')
      }

      // @vue-computed tag support.
      if (item._vueComputed) {
        // Strip out HTML table from description.
        item.description = item.description.replace(/<\/p.*/ms, '')

        item._vueComputed.map(property => {
          templateData.push({
            id: `module:${data.displayName}.computed.${property.name}`,
            longname: `module:${data.displayName}.computed.${property.name}`,
            kind: 'property',
            scope: 'static',
            memberof: `module:${data.displayName}.computed`,

            ...property
          })
        })
      }

      // Vue.js Computed properties fixes.
      if (parts[1] === 'computed' && parts.length === 2) {
        item.kind = 'property'
      }
    })
  }

  /**
   * Write template data to file system.
   *
   * @param {string} file - The file name.
   * @param {*} data - The JSDoc / Vuedoc generated data.
   */
  writeTemplateData(file, templateData) {
    if (!templateData) return

    const content = dmd(templateData, {
      'example-lang': 'vue',
      'heading-depth': 2,
      partial: [
        path.resolve(cwd, 'partials/main.hbs'),
        path.resolve(cwd, 'partials/sig-name.hbs')
      ],
      separators: true
    })
    if (!content) return

    const destination = `${this.destination}/api/${file.replace('src/', '').replace(/\.[^/.]+$/, '.md')}`
    mkdirp.sync(path.dirname(destination))

    // Build frontmatter.
    const name = templateData[0].id.split(':').pop()
    const frontmatter = `---\ntitle: ${name}\n---\n\n`

    fs.writeFileSync(destination, frontmatter + content)
  }
}

export { DruxtDocgen }
