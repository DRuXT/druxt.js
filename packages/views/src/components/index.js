/**
 * The `<DruxtView />` Vue.js component.
 *
 * @type {object}
 * @exports DruxtView
 * @see {@link ./DruxtView|DruxtView}
 */
export { default as DruxtView } from './DruxtView.vue'

/**
 * The `<DruxtViewsFilter />` Vue.js component.
 *
 * @type {object}
 * @exports DruxtViewsFilter
 * @see {@link ./DruxtViewsFilter|DruxtViewsFilter}
 */
export { default as DruxtViewsFilter } from './DruxtViewsFilter.vue'

/**
 * The `<DruxtViewsFilters />` Vue.js component.
 *
 * @type {object}
 * @exports DruxtViewsFilters
 * @see {@link ./DruxtViewsFilters|DruxtViewsFilters}
 */
export { default as DruxtViewsFilters} from './DruxtViewsFilters.vue'

/**
 * The `<DruxtViewsPager />` Vue.js component.
 *
 * @type {object}
 * @exports DruxtViewsPager
 * @see {@link ./DruxtViewsPager|DruxtViewsPager}
 */
export { default as DruxtViewsPager } from './DruxtViewsPager.vue'

/**
 * The `<DruxtViewsSorts />` Vue.js component.
 *
 * @type {object}
 * @exports DruxtViewsSorts
 * @see {@link ./DruxtViewsSorts|DruxtViewsSorts}
 */
export { default as DruxtViewsSorts } from './DruxtViewsSorts.vue'

/**
 * The `<DruxtRouterView />` Vue.js component.
 *
 * @type {object}
 * @exports DruxtRouterView
 * @see {@link ./DruxtRouterView|DruxtRouterView}
 */
export { default as DruxtRouterView} from './DruxtRouterView.vue'

export * from './blocks'
