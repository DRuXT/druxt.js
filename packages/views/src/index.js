/**
 * Vuex store module.
 *
 * @type {object}
 * @exports DruxtViewsStore
 * @name DruxtViewsStore
 * @see {@link ./stores/views|DruxtViewsStore}
 */
export { DruxtViewsStore } from './stores/views'

/**
 * Vue.js mixins.
 *
 * @type {object}
 * @exports DruxtViewsMixins
 * @see {@link ./mixins|DruxtViewsMixins}
 * @see {@link ./mixins/filter|DruxtViewsFilterMixin}
 * @see {@link ./mixins/filters|DruxtViewsFiltersMixin}
 * @see {@link ./mixins/pager|DruxtViewsPagerMixin}
 * @see {@link ./mixins/sorts|DruxtViewsSortsMixin}
 * @see {@link ./mixins/view|DruxtViewsViewMixin}
 */
export * from './mixins'

/**
 * Default function to alert user to incorrectly installed module.
 *
 * This was added as part of the @nuxt/kit update due to breaking changes.
 */
export default () => {
  throw new Error("DruxtViews Nuxt module must be installed as 'druxt-views/nuxt'")
}
