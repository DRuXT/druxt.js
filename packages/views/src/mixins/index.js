/**
 * Vue.js mixin for DruxtViewsFilter slot themable components.
 *
 * @type {object}
 * @exports DruxtViewsFilterMixin
 * @see {@link ./filter|DruxtViewsFilterMixin}
 */
export { DruxtViewsFilterMixin } from './filter'

/**
 * Vue.js mixin for DruxtViewsFilters slot themable components.
 *
 * @type {object}
 * @exports DruxtViewsFiltersMixin
 * @see {@link ./filters|DruxtViewsFiltersMixin}
 */
export { DruxtViewsFiltersMixin } from './filters'

/**
 * Vue.js mixin for DruxtViewsPager slot themable components.
 *
 * @type {object}
 * @exports DruxtViewsPagerMixin
 * @see {@link ./pager|DruxtViewsPagerMixin}
 */
export { DruxtViewsPagerMixin } from './pager'

/**
 * Vue.js mixin for DruxtViewSorts slot themable components
 *
 * @type {object}
 * @exports DruxtViewsSortsMixin
 * @see {@link ./sorts|DruxtViewsSortsMixin}
 */
export { DruxtViewsSortsMixin } from './sorts'

/**
 * Vue.js mixin for DruxtView slot themable components
 *
 * @type {object}
 * @exports DruxtViewsViewMixin
 * @see {@link ./view|DruxtViewsViewMixin}
 */
export { DruxtViewsViewMixin } from './view'
