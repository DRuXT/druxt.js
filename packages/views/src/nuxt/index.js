import { addPluginTemplate, defineNuxtModule, installModule } from '@nuxt/kit'
import { join, resolve } from 'path'
import DruxtViewsStorybook from './storybook'

/**
 * The Nuxt.js module function.
 *
 * - Adds the Vue.js components to the Nuxt application.
 *
 * The module function should not be used directly, but rather installed via your Nuxt configuration file.
 *
 * Options are set on the root level `druxt` Nuxt config object.
 *
 * @example @lang js
 * // `nuxt.config.js`
 * module.exports = {
 *   modules: [
 *     'druxt-views'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 *
 * @param {object} moduleOptions - Nuxt module options object.
 */
const DruxtViewsNuxtModule = defineNuxtModule({
  meta: {
    name: 'druxt-views',
  },
  defaults: {
    baseUrl: '',
    endpoint: '/jsonapi'
  },

  async setup(moduleOptions, nuxt) {
    // Set default options.
    const options = {
      baseUrl: moduleOptions.baseUrl,
      ...nuxt.options?.druxt || {},
      views: {
        query: {},
        ...nuxt.options?.druxt?.views,
        ...moduleOptions,
      }
    }

    // Register components directories.
    nuxt.hook('components:dirs', dirs => {
      dirs.push({ path: join(__dirname, '../dist/components') })
      dirs.push({ path: join(__dirname, '../dist/components/blocks') })
    })

    // Add dependant modules.
    await installModule('druxt/nuxt', options, nuxt)
    const modules = ['druxt-entity/nuxt', 'druxt-schema/nuxt']
    for (const module of modules) {
      await installModule(module, { baseUrl: options.baseUrl }, nuxt)
    }

    // Add Vuex plugin.
    addPluginTemplate({
      src: resolve(__dirname, '../templates/store.js'),
      fileName: 'store/druxt-views.js',
      options: options.druxt
    })

    // Enable Vuex Store.
    nuxt.options.store = true

    // Nuxt Storybook.
    nuxt.hook('storybook:config', async ({ stories }) => {
      await DruxtViewsStorybook.call(nuxt, { stories })
    })
  }
})

export default DruxtViewsNuxtModule
