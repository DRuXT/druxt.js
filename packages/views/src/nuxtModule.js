import { resolve } from 'path'
import DruxtViewsStorybook from './nuxtStorybook'

/**
 * The Nuxt.js module function.
 *
 * - Adds the Vue.js components to the Nuxt application.
 *
 * The module function should not be used directly, but rather installed via your Nuxt configuration file.
 *
 * Options are set on the root level `druxt` Nuxt config object.
 *
 * @example @lang js
 * // `nuxt.config.js`
 * module.exports = {
 *   modules: [
 *     'druxt-views'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 *
 * @param {object} moduleOptions - Nuxt module options object.
 */
const DruxtViewsNuxtModule = function (moduleOptions = {}) {
  const { options } = this

  // Use root level Druxt options.
  if (typeof this.options === 'undefined' || !options.druxt) {
    throw new TypeError('Druxt settings missing.')
  }

  // Add dependant modules.
  const modules = ['druxt', 'druxt-entity', 'druxt-schema'].filter((module) => !options.modules.includes(module))
  for (const module of modules) {
    this.addModule(module)
  }

  // Add plugin.
  this.addPlugin({
    src: resolve(__dirname, '../nuxt/plugin.js'),
    fileName: 'druxt-views.js',
    options: options.druxt
  })

  // Add Vuex plugin.
  this.addPlugin({
    src: resolve(__dirname, '../nuxt/store.js'),
    fileName: 'store/druxt-views.js',
    options: options.druxt
  })

  // Enable Vuex Store.
  options.store = true

  // Nuxt Storybook.
  this.nuxt.hook('storybook:config', async ({ stories }) => {
    await DruxtViewsStorybook.call(this, { stories })
  })
}

export { DruxtViewsNuxtModule }
