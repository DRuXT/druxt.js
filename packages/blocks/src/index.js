import { DruxtBlocksNuxtModule } from './nuxtModule'

/**
 * The Nuxt.js module function.
 *
 * Installs the module functionality in a Nuxt application.
 *
 * @type {Function}
 * @exports default
 * @name DruxtBlocksModule
 * @see {@link ./nuxtModule|DruxtBlocksModule}
 *
 * @example <caption>nuxt.config.js</caption> @lang js
 * module.exports = {
 *   modules: [
 *     'druxt-blocks'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 */
export default DruxtBlocksNuxtModule

/**
 * Vue.js Mixin.
 *
 * @type {object}
 * @exports DruxtBlocksBlockMixin
 * @see {@link ./mixins/block|DruxtBlocksBlockMixin}
 */
 export { DruxtBlocksBlockMixin } from './mixins/block'
