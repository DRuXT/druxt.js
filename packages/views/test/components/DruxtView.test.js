import 'regenerator-runtime/runtime'
import { createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'

import { DruxtClient, DruxtStore } from '../../../druxt/src'
import { DruxtViewsStore } from '../../src'
import DruxtView from '../../src/components/DruxtView.vue'

// Setup local vue instance.
const localVue = createLocalVue()
localVue.use(Vuex)

// Setup mock Vuex store.
let store

const mocks = {
  $createElement: jest.fn(),
  $druxt: {
    settings: {}
  },
  $fetch: jest.fn(),
  $fetchState: {
    pending: true
  },
  $route: {
    query: {}
  }
}

const stubs = ['DruxtEntity']
const mountComponent = (propsData) => mount(DruxtView, { localVue, mocks, propsData, store, stubs })

describe('DruxtView', () => {
  beforeEach(() => {
    store = new Vuex.Store()

    DruxtStore({ store })
    store.$druxt = new DruxtClient('https://demo-api.druxtjs.org')

    DruxtViewsStore({ store })

    store.app = { context: { error: jest.fn() }, store }
  })

  test('featured_articles', async () => {
    const wrapper = mountComponent({
      displayId: 'page_1',
      viewId: 'featured_articles'
    })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Fetch key.
    expect(DruxtView.fetchKey.call(wrapper.vm, jest.fn(() => 0))).toBe('DruxtView:featured_articles:page_1:0')

    // Props.
    expect(wrapper.vm.displayId).toBe('page_1')
    expect(wrapper.vm.uuid).toBe(null)
    expect(wrapper.vm.viewId).toBe('featured_articles')

    // Computed.
    expect(wrapper.vm.attachments_after).toStrictEqual([])
    expect(wrapper.vm.attachments_before).toStrictEqual([])
    expect(wrapper.vm.filters).toStrictEqual([])
    expect(wrapper.vm.headers).toStrictEqual([])
    expect(wrapper.vm.mode).toBe('card')

    // Slots.
    expect(Object.keys(wrapper.vm.getScopedSlots())).toStrictEqual([
      'header',
      'attachments_before',
      'results',
      'pager',
      'attachments_after',
      'default'
    ])

    const h = jest.fn()
    const slotsMock = {
      $createElement: h,
      display: wrapper.vm.display,
      filters: wrapper.vm.filters,
      results: wrapper.vm.results,
    }
    expect(DruxtView.druxt.slots.call(slotsMock, h).results({}).length).toBe(8)
    expect(h).toHaveBeenCalledTimes(8)
    expect(h).toHaveBeenCalledWith('DruxtEntity', expect.any(Object))

    // Test empty results.
    slotsMock.display.display_options.empty = [{
      content: '<p>No content</p>',
      plugin_id: 'text_custom',
    }]
    slotsMock.results = []
    expect(DruxtView.druxt.slots.call(slotsMock, h).results({}).length).toBe(1)
    expect(h).toHaveBeenCalledTimes(9)
    expect(h).toHaveBeenCalledWith('div', expect.any(Object))

    // DruxtModule.
    expect(wrapper.vm.component.is).toBe('DruxtWrapper')
    expect(wrapper.vm.component.options).toStrictEqual([
      'DruxtViewFeaturedArticlesPage1',
      'DruxtViewFeaturedArticles'
    ])

    // Query.
    await wrapper.setProps({ arguments: [1, 2, 3] })
    wrapper.vm.model = {
      page: 1,
      filter: { test: 1 },
      sort: 'test'
    }
    expect(wrapper.vm.getQuery(wrapper.vm.component.settings)).toStrictEqual({
      page: 1,
      'views-argument[0]': 1,
      'views-argument[1]': 2,
      'views-argument[2]': 3,
      'views-filter': { test: 1 },
      'views-sort[sort_by]': 'test',
    })
    expect(wrapper.vm.showPager).toBe(true)
    expect(wrapper.vm.showSorts).toBe(false)

    wrapper.vm.onFiltersUpdate()
    expect(wrapper.vm.model.page).toBe(null)
    expect(wrapper.vm.model.sort).toBe(null)

    wrapper.vm.$route.query = {}
    await wrapper.setProps({ arguments: [] })
    await localVue.nextTick()

    expect(wrapper.vm.model).toStrictEqual({
      filter: {},
      page: null,
      sort: null
    })
    expect(wrapper.vm.getQuery(wrapper.vm.component.settings)).toStrictEqual({})

    // Watches.
    expect(mocks.$fetch).toHaveBeenCalledTimes(0)
    await DruxtView.watch.displayId.call(mocks)
    expect(mocks.$fetch).toHaveBeenCalledTimes(1)
    wrapper.vm.model.page = 1
    await localVue.nextTick()
    expect(mocks.$fetch).toHaveBeenCalledTimes(2)
    wrapper.vm.model.sort = 'nid'
    await localVue.nextTick()
    expect(mocks.$fetch).toHaveBeenCalledTimes(3)
    await DruxtView.watch.query.call(mocks)
    expect(mocks.$fetch).toHaveBeenCalledTimes(4)
    await DruxtView.watch.uuid.call(mocks)
    expect(mocks.$fetch).toHaveBeenCalledTimes(5)
    await DruxtView.watch.viewId.call(mocks)
    expect(mocks.$fetch).toHaveBeenCalledTimes(6)
  })

  // TODO : Fix test for filters / attachments, etc.
  // test('recipes--embed', async () => {
  //   const wrapper = mountComponent({
  //     displayId: 'embed_1',
  //     uuid: 'f6c38097-d534-4bfb-87d9-09526fe44e9c',
  //   })
  //   await wrapper.vm.$options.fetch.call(wrapper.vm)

  //   // Fetch key.
  //   expect(DruxtView.fetchKey.call(wrapper.vm, jest.fn(() => 0))).toBe('DruxtView:f6c38097-d534-4bfb-87d9-09526fe44e9c:embed_1:0')

  //   // Scoped slots.
  //   const scopedSlots = wrapper.vm.getScopedSlots.call(mocks)
  //   expect(Object.keys(scopedSlots)).toStrictEqual([
  //     'header',
  //     // 'filters',
  //     // 'sorts',
  //     // 'attachments_before',
  //     'results',
  //     // 'pager',
  //     // 'attachments_after',
  //     'default'
  //   ])

  //   const h = jest.fn()
  //   const mockSlots = DruxtView.druxt.slots.call({
  //     $createElement: h,
  //     headers: wrapper.vm.headers,
  //     filters: wrapper.vm.filters,
  //     results: wrapper.vm.results,
  //   }, h)
  //   for (const key of Object.keys(mockSlots).filter((key) => key !== 'default')) {
  //     mockSlots[key] = jest.fn()
  //   }
  //   mockSlots.default({ foo: 'bar' })
  //   expect(mockSlots.header).toBeCalledWith({ foo: 'bar' })
  //   expect(mockSlots.filters).toBeCalledWith({ foo: 'bar' })
  //   expect(mockSlots.results).toBeCalledWith({ foo: 'bar' })
  // })

  test('getQuery', async () => {
    const mock = { arguments: [], model: {} }
    expect(DruxtView.methods.getQuery.call(mock)).toStrictEqual({})

    mock.display = { display_options: { filters: [{
      entity_type: 'node',
      plugin_id: 'bundle',
      value: { page: 'page' }
    }] } }
    expect(DruxtView.methods.getQuery.call(mock)).toStrictEqual({})

    // Druxt query settings.
    const settings = { query: { fields: ['title'] }}
    expect(DruxtView.methods.getQuery.call(mock, settings)).toStrictEqual({})

    settings.query.bundleFilter = true
    expect(DruxtView.methods.getQuery.call(mock, settings)).toStrictEqual({
      'fields[node--page]': 'uuid,title'
    })
  })

  test('v-model', () => {
    const h = jest.fn((tag, options, children) => {
      options.on.input.call(mock, options.attrs.mock)
    })
    const mock = {
      filters: [{
        id: 'test',
        plugin_id: 'test',
        expose: {
          identifier: 'test',
        },
      }],
      model: {
        filter: {},
        page: null,
        sort: null,
      },
      results: [],
      showPager: true,
      showSorts: true,
    }
    const slots = DruxtView.druxt.slots.call(mock, h)

    // Filters
    expect(mock.model.filter).toStrictEqual({})
    slots.filters({ mock: { test: 1 } })
    expect(mock.model.filter).toStrictEqual({ test: 1 })

    // Sorts
    expect(mock.model.sort).toStrictEqual(null)
    slots.sorts({ mock: 'test' })
    expect(mock.model.sort).toStrictEqual('test')

    // Pager
    expect(mock.model.page).toStrictEqual(null)
    slots.pager({ mock: 1 })
    expect(mock.model.page).toStrictEqual(1)
  })
})
