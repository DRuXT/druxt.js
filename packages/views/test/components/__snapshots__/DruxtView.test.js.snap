// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DruxtView featured_articles 1`] = `
Array [
  "DruxtViewFeaturedArticlesPage1",
  "DruxtViewA4cd303b354943f99868886099717e50Page1",
  "DruxtViewFeaturedArticles",
  "DruxtViewA4cd303b354943f99868886099717e50",
  "DruxtViewPage1",
]
`;
